[IMP] *: translation methods cleaning

mart-e opened the pull request:
Extract some cleaning wanted in #30759 but not related to language packs

- `amount_to_text` should work only on activated languages
- deprecate `load_lang` and rely on `_create_lang` or `_activate_lang`
- `trans_load_data` and `_load_module_terms` no longer silently activate languages
- pass explicitly parameters instead of relying on context content
- replace some `IrTranslation._load_module_terms(['base'], ['fr_FR'])` by `BaseModule._update_translations(['fr_FR'])` for higher level methods
- create `TranslationExporter` class to clarify the `trans_export` method (and clean dead code)

Task-id: 2088290
Pad: https://pad.odoo.com/p/r.f6f789f8711d21314bd902972153ea9d

xmo-odoo commented on odoo/addons/base/models/res_lang.py:
```diff
@@ -88,8 +88,24 @@ def _register_hook(self):
+
+    def _activate_lang(self, code=False):
+        """ Activate languages
+        :param code: optional code of the language to activate, using self if not specified
```
That's a super weird API, I'd rather the method either always ignored its subject (and only used the code) *or* always used the code and there was a version of _lang_get (or even a modification on _lang_get itself) which allowed getting deactivated langs.

xmo-odoo commented on odoo/addons/base/models/res_lang.py:
```diff
@@ -88,8 +88,24 @@ def _register_hook(self):
             _logger.error("No language is active.")
 
-    @api.model
+    # TODO v14, remove me
```
That's todo v15 (since the current master is the future v14).

Also nobody's going to check this anyway. If you want that done, put an assertion based on odoo.release.version_info.

mart-e replied:
technically, master is `13.0alpha1`, I don't really care if we remove the method in 2 years or 2 weeks, just want to make sure nobody else is relying on it before removing it (warning should allow to detect that on the runbot)

xmo-odoo replied:
@mart-e it should be 14a1, @KangOl just forgot to bump the version of master after release.

See #38913

rco-odoo commented on odoo/addons/base/models/ir_translation.py:
```diff
@@ -27,15 +27,15 @@ class IrTranslationImport(object):
     _table = 'tmp_ir_translation_import'
 
-    def __init__(self, model):
+    def __init__(self, cr, table, overwrite=False):
```
I never understood the point of being generic here.
The value of `table`  is *always* `'ir_translation'`.
I would simply remove it from the parameters.
```suggestion
    def __init__(self, cr, overwrite=False):
```

rco-odoo commented on odoo/addons/base/models/res_lang.py:
```diff
@@ -88,15 +88,24 @@ def _register_hook(self):
+        :return: the language matching 'code' activated
+        """
+        lang = self.with_context(active_test=False).search([('code', '=', code)])
+        lang.filtered(lambda l: not l.active).write({'active': True})
```
The `code` is unique, so by construction `lang` contains at most 1 record.
What about a plain stupid `if` statement like
```suggestion
        if lang and not lang.active:
            lang.active = True
```

rco-odoo commented on odoo/addons/base/models/res_lang.py:
```diff
@@ -157,18 +166,18 @@ def fix_datetime_format(format):
         # config['load_language'] is a comma-separated list or None
         lang_code = (tools.config.get('load_language') or 'en_US').split(',')[0]
-        lang = self._lang_get(lang_code)
+        lang = self._activate_lang(lang_code)
```
Well, your API suggests a very nice idiom here:
```suggestion
        lang = self._activate_lang(lang_code) or self._create_lang(lang_code)
```

rco-odoo commented on odoo/addons/base/models/res_lang.py:
```diff
@@ -88,15 +88,24 @@ def _register_hook(self):
+        lang.filtered(lambda l: not l.active).write({'active': True})
+        return lang
 
+    def _create_lang(self, lang, lang_name=None):
```
Maybe make it return the same thing as `_activate_lang`.

rco-odoo commented on odoo/addons/base/wizard/base_import_language.py:
```diff
@@ -37,8 +36,11 @@ def import_lang(self):
 
-                tools.trans_load_data(this._cr, buf, fileformat, this.code,
-                                      lang_name=this.name, context=this._context)
+                lang = self.env['res.lang']._activate_lang(self.code)
```
Again, the idiom:
```suggestion
                Lang = self.env['res.lang']
                lang = Lang._activate_lang(self.code) or Lang._create_lang(self.code, self.name)
```

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -997,76 +1003,110 @@ def babel_extract_terms(fname, path, root, extract_method="python", trans_type='
+        return out
 
 
 def trans_load(cr, filename, lang, verbose=True, module_name=None, context=None):
```
What about making parameters `create_empty_translation` and `overwrite` explicit here, too?

mart-e commented:
Added explicit parameters to `trans_load` too and removed the parameter `module_name` which was misleading (only used for code translation) and should be deductible from po file content

rco-odoo commented on odoo/addons/base/models/ir_module.py:
```diff
@@ -886,7 +886,7 @@ def _update_translations(self, filter_lang=None):
         }
         mod_names = topological_sort(mod_dict)
-        self.env['ir.translation']._load_module_terms(mod_names, filter_lang)
+        self.env['ir.translation']._load_module_terms(mod_names, filter_lang, overwrite=overwrite)
```
```suggestion
        self.env['ir.translation']._load_module_terms(mod_names, filter_lang, overwrite)
```

rco-odoo commented on odoo/addons/base/models/ir_translation.py:
```diff
@@ -774,24 +774,18 @@ def make_domain(fld, rec):
+    def _get_import_cursor(self, overwrite):
         """ Return a cursor-like object for fast inserting translations """
-        return IrTranslationImport(self)
+        return IrTranslationImport(self._cr, overwrite=overwrite)
```
```suggestion
        return IrTranslationImport(self._cr, overwrite)
```

rco-odoo commented on odoo/addons/base/wizard/base_language_install.py:
```diff
@@ -33,7 +33,8 @@ def _get_languages(self):
         mods = self.env['ir.module.module'].search([('state', '=', 'installed')])
-        mods.with_context(overwrite=self.overwrite)._update_translations(self.lang)
+        self.env['res.lang']._activate_lang(self.lang)
+        mods._update_translations(self.lang, overwrite=self.overwrite)
```
```suggestion
        mods._update_translations(self.lang, self.overwrite)
```

rco-odoo commented on odoo/addons/test_translation_import/tests/test_term_count.py:
```diff
@@ -259,10 +268,12 @@ def test_export_import(self):
             ('lang', '=', 'fr_FR'),
-            ('module', '=', 'test_translation_import')
+            ('module', '=', 'test_translation_import'),
+            ('value', '!=', '')
```
```suggestion
            ('value', '!=', ''),
```

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -997,76 +1003,106 @@ def babel_extract_terms(fname, path, root, extract_method="python", trans_type='
-        irt_cursor = Translation._get_import_cursor()
+        # read the rest of the file with a cursor-like object for fast inserting translations"
+        Translation = env['ir.translation']
+        irt_cursor = Translation._get_import_cursor(overwrite=overwrite)
```
```suggestion
        irt_cursor = Translation._get_import_cursor(overwrite)
```

rco-odoo commented on odoo/addons/base/tests/test_qweb.py:
```diff
@@ -522,10 +522,9 @@ def test_render_t_call_propagates_t_lang(self):
 
-        self.env['res.lang']._lang_get(other_lang).write({
-            'active': True,
+        other.write({
```
Naming: maybe use `language` instead of `other`...

rco-odoo commented on odoo/addons/test_translation_import/tests/test_term_count.py:
```diff
@@ -232,7 +240,8 @@ def test_export_import(self):
             ('lang', '=', 'fr_FR'),
-            ('module', '=', 'test_translation_import')
+            ('module', '=', 'test_translation_import'),
+            ('value', '!=', '')
```
```suggestion
            ('value', '!=', ''),
```

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -997,76 +1003,106 @@ def babel_extract_terms(fname, path, root, extract_method="python", trans_type='
-
-def trans_load(cr, filename, lang, verbose=True, module_name=None, context=None):
+    def export_translatable_resources(self):
+        """ Export translations for static resources """
```
Docstring: can you give a hint on what are static resources?
Is this code translations?

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -997,76 +1003,106 @@ def babel_extract_terms(fname, path, root, extract_method="python", trans_type='
+                    # due to topdown, first iteration is in first level
+                    break
+
+    def read_rows(self, lang=None):
```
The name is confusing, as this object collects data for exportation.
I suggest `get_rows` (you retrieve the rows from the exporter object.)

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -785,7 +785,11 @@ def write_rows(self, rows):
+    reader = TranslationExporter(cr, modules=modules)
+    reader.export_translatable_records()
+    reader.export_translatable_resources()
+    translations = reader.read_rows(lang)
```
The `TranslationExporter` just makes the API more confusing from a user's point of view:
- this is not a "reader" in the usual meaning of the term;
- the methods `export_translatable_records` and `export_translatable_resources` look cryptic:
   - should they be public?
   - does it make any sense to not call both of them?

I suggest the following alternative API:
```python
translations = TranslationExporter(cr, modules).get_rows(lang)
```

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -785,7 +785,11 @@ def write_rows(self, rows):
+    reader.export_translatable_resources()
+    translations = reader.read_rows(lang)
+
     modules = set(t[0] for t in translations)
```
By the way, what is the point of this statement?
It should give the same module names as the ones passed in argument, shouldn't it?

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -887,107 +906,94 @@ def push_translation(module, type, name, id, source, comments=None):
 
-    for (xml_name, model, res_id, module) in cr.fetchall():
-        xml_name = "%s.%s" % (module, xml_name)
+        for (xml_name, model, res_id, module) in self._cr.fetchall():
```
Note for later: we should group records by `model`, and fetch their translations in batch.

xmo-odoo commented on odoo/tools/translate.py:
```diff
@@ -887,107 +917,94 @@ def push_translation(module, type, name, id, source, comments=None):
+                if os.path.sep != '/':
+                    display_path = display_path.replace(os.path.sep, '/')
+                return module, fabsolutepath, frelativepath, display_path
+            return None, None, None, None
```
Why keep these as nested functions instead of making them regular methods?

xmo-odoo commented on odoo/tools/translate.py:
```diff
@@ -997,76 +1014,92 @@ def babel_extract_terms(fname, path, root, extract_method="python", trans_type='
                 # Babel 0.9.6 yields lineno, message, comments
                 # Babel 1.3 yields lineno, message, comments, context
                 lineno, message, comments = extracted[:3]
```
Could at least remove the comment given we now have a hard dependency on 2.3.

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -886,107 +914,152 @@ def push_translation(module, type, name, id, source, comments=None):
+
+        if 'all' not in self._modules:
+            query = query.format(module_clause="module = ANY(%s)")
+            query_param = self._modules
```
The collection `self._modules` is an input, and not guaranteed to be a list
```suggestion
            query_param = list(self._modules)
```

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -886,107 +914,152 @@ def push_translation(module, type, name, id, source, comments=None):
+
+        query = """SELECT min(name), model, res_id, module
+                     FROM ir_model_data
+                    WHERE {module_clause}
```
It is always the same `module_clause`, so
```suggestion
                    WHERE module = ANY(%s)
```

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -886,107 +914,152 @@ def push_translation(module, type, name, id, source, comments=None):
-        if not translatable_model(record):
-            continue
+        records_per_model = defaultdict(dict)
+        imd_info = namedtuple('ExternalId', ['name', 'model', 'res_id', 'module'])
```
You should make this namedtuple definition available globally.
This would ease the understanding of `_get_translatable_records`, for instance.

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -886,107 +914,152 @@ def push_translation(module, type, name, id, source, comments=None):
 
-    query += ' GROUP BY model, res_id, module ORDER BY module, model, min(name)'
+        if not self.env[model]._translate:
+            return False
```
The method should always return a recordset, possibly empty.
This would avoid the necessary
```python
if not records:
    continue
```
in the caller's loop below.

mart-e commented:
@robodoo r+ merge

robodoo commented:
Merge method set to merge directly, using the PR as merge commit message

mart-e commented:
robodoo r+ again, my bad

robodoo commented:
Linked pull request(s) odoo/enterprise#6210 not ready. Linked PRs are not staged until all of them are ready.

robodoo commented:
Merged at f591ec1ea992b624657c53edc2744f1663d09f28, thanks!
